### PR TITLE
[core] Reconstruction for lost plasma objects

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Automatically attempt to recover objects that were stored in plasma and lost due to a failure. The logic for recovery is encapsulated in the `ObjectRecoveryManager`. The algorithm works as follows:

1. Check that the object is missing from the direct memory store and that
we own the object. If either is false, then fail the recovery operation.
2. Look up the object in the global directory to check for other
locations of the object. If another location exists, attempt to pin it.
If the pinning is successful, then mark the recovery as a success by
storing a new value for the object in the direct memory store.
3. If pinning fails at all locations for the object (or there are no
locations), attempt to reconstruct the object by resubmitting the task
that created the object. If the task resubmission fails, then the
fail the recovery operation.
4. If task resubmission succeeds, recursively attempt to recover any
plasma arguments to the task. The recovery operation will succeed once
the task completes and stores a new value for its return object.

Other changes in the PR include:
1. The owner keeps track of where objects are pinned.
2. The `TaskManager` and `ReferenceCounter` allow tasks to be resubmitted.

Recovery through task resubmission (along with lineage pinning from #7690) is off by default. It can be enabled with an internal config parameter:
```
ray.init(_internal_config=json.dumps({
    "lineage_pinning_enabled": 1,
  }))
```

This PR only covers reconstruction of objects where all tasks in the object's lineage were submitted by one process. Reconstruction for borrowed objects is left for a future PR. This is more complicated to do because we'd have to modify the lineage ref counting protocol to include borrowers' ref counts.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by stephanie-wang** [comment]
@ericl and @edoakes, this PR is big (although a lot of it is test cases). If it's too large to review, let me know and I can split it into one PR that just attempts to pin a new copy and a second to reconstruct objects through task resubmission.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc
Comment (edoakes) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Oh, this answers my question before. We're now removing the entries when the node fails. Got it.

**Review by edoakes — COMMENTED** [review]
Didn't have a chance to read the tests closely, but the main logic looks pretty straightforward. It seems like we're putting a lot more reliance back on the GCS which is a bit disappointing, but not sure if it can be avoided.

Would it be possible to somehow have owners directly detect failures of raylets that are pinning their owned objects by modifying the existing `WaitForObjectEviction` RPC protocol? I guess we really want a bidirectional signal - the owner and raylet each know if the other goes down.

File: src/ray/core_worker/object_recovery_manager.cc
Comment (stephanie-wang) [reply src/ray/core_worker/object_recovery_manager.cc:26]: It's only used internally, but sometimes we want to RAY_CHECK it and sometimes this is a valid error (e.g., if you try to re-execute a task that depends on a borrowed object).

File: src/ray/core_worker/object_recovery_manager.h
Comment (stephanie-wang) [reply src/ray/core_worker/object_recovery_manager.h:71]: Uhh I'm not sure how to answer this question. This class just needs some sort of "global lookup" function. Whether that checks some local data structures or the GCS or something else doesn't really matter.

File: src/ray/core_worker/object_recovery_manager.h
Comment (stephanie-wang) [reply src/ray/core_worker/object_recovery_manager.h:79]: It's nice to resubmit the task first because then we can be sure that the reconstructing objects will get pinned by their ref counts. Maybe this is okay even in the reverse order, but it seemed more complicated to convince myself of the correctness.

File: src/ray/core_worker/object_recovery_manager.h
Comment (stephanie-wang) [reply src/ray/core_worker/object_recovery_manager.h:140]: Yeah :) I can do it after this PR is merged.

**Comment by stephanie-wang** [comment]
> Didn't have a chance to read the tests closely, but the main logic looks pretty straightforward. It seems like we're putting a lot more reliance back on the GCS which is a bit disappointing, but not sure if it can be avoided.
> 
> Would it be possible to somehow have owners directly detect failures of raylets that are pinning their owned objects by modifying the existing `WaitForObjectEviction` RPC protocol? I guess we really want a bidirectional signal - the owner and raylet each know if the other goes down.

Still need to update with your comments, but just to comment on the GCS thing. I think that reliance on the GCS is actually fine as long as we are smart about it, i.e., the interface is very narrow. In some cases, the right GCS interface can really simplify the worker logic. I do believe very strongly that the GCS should contain the information about which nodes are currently in the cluster. It's pretty standard to store this sort of configuration information in a replicated and centralized service, since it makes it simple to keep it consistent and fault-tolerant.

File: src/ray/common/ray_object.h
Comment (zhijunfu) [inline src/ray/common/ray_object.h:113]: Since we're already use `meta` to indicate if the object is in plasma, should we reuse `data` for the optional `raylet id` in this case?

File: src/ray/core_worker/core_worker.cc
Comment (zhijunfu) [inline src/ray/core_worker/core_worker.cc:410]: There could be quite a lot of objects on a given node, how about using `DEBUG` here?

File: src/ray/core_worker/object_recovery_manager.h
Comment (zhijunfu) [inline src/ray/core_worker/object_recovery_manager.h:36]: I assume this would only recover objects from common tasks, not actor tasks?

**Review by edoakes — COMMENTED** [review]
LGTM but I'll let @ericl approve so he has a chance to review before merging.

File: src/ray/common/ray_object.h
Comment (stephanie-wang) [reply src/ray/common/ray_object.h:113]: Hmm we could, but I don't think this is such a good idea because the raylet ID is a pretty different concept from the object value. I'd rather just keep the concepts separate so it's clearer which field to check.

File: src/ray/core_worker/object_recovery_manager.h
Comment (stephanie-wang) [reply src/ray/core_worker/object_recovery_manager.h:36]: Yes, it actually does not differentiate between actor and non-actor tasks, but by default actor tasks won't be reconstructed because they are added to the `TaskManager` with 0 maximum retries.

File: src/ray/core_worker/core_worker.cc
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc:410]: Actually, I think this is probably okay because it's `INFO`, not `WARNING`, so it won't go to the driver. It's important to have this information in the backend logs by default.

File: src/ray/common/ray_object.h
Comment (ericl) [inline src/ray/common/ray_object.h:70]: I find it weird that the raylet id (which seems like a global runtime property) is kept as an attribute of RayObject. Doesn't it make more sense to query some global registry to check where the object is pinned?

File: src/ray/core_worker/core_worker.h
Comment (ericl) [inline src/ray/core_worker/core_worker.h:795]: Add a comment.

File: src/ray/core_worker/object_recovery_manager.h
Comment (ericl) [inline src/ray/core_worker/object_recovery_manager.h:87]: Does this recursive recovery issue RPCs to their owners?

File: src/ray/core_worker/object_recovery_manager.h
Comment (ericl) [inline src/ray/core_worker/object_recovery_manager.h:70]: Why not delegate non-owned object recovery to the owner instead of failing? Seems like we have to do that for recursive arg dependencies anyways.

File: src/ray/raylet/raylet_client.h
Comment (ericl) [inline src/ray/raylet/raylet_client.h:60]: Hmm why not use gRPC for both?

**Comment by ericl** [comment]
I kind of expected at least a couple new proto messages for reconstruction, are they not needed?

File: src/ray/raylet/raylet_client.h
Comment (stephanie-wang) [reply src/ray/raylet/raylet_client.h:60]: Er I'm not sure, I was just reusing what was already here.

File: src/ray/core_worker/object_recovery_manager.h
Comment (stephanie-wang) [reply src/ray/core_worker/object_recovery_manager.h:70]: I think we should do it, but I thought it'd be better to save that for a future PR, just to keep the changes incremental. This PR only covers reconstruction of objects where the object's lineage is all owned by one process. Reconstruction for borrowed objects is more complicated because we'd have to modify the lineage ref counting protocol to include borrowers' ref counts. I'll update the PR description to account for this.

File: src/ray/common/ray_object.h
Comment (stephanie-wang) [reply src/ray/common/ray_object.h:70]: Yeah I actually tried to do this at first in the `ObjectRecoveryManager`, but it turned out into a bit of a headache trying keep it in sync with the ref counter correctly (e.g., we don't want to set the raylet ID if the ref count is already 0). It's nice to keep it in the RayObject because then we can maintain this invariant easily: A plasma object that we own can be fetched <=> the raylet location in the RayObject is set.

**Comment by stephanie-wang** [comment]
> I kind of expected at least a couple new proto messages for reconstruction, are they not needed?

I think it will only be needed if object IDs escape. I guess once the GCS service is finished, we will also need it to reconstruct arguments to an actor constructor task.

File: src/ray/common/ray_object.h
Comment (ericl) [reply src/ray/common/ray_object.h:70]: Not sure I follow. I thought RayObject was a "dummy struct" class, and the pinned location can change. If it changes wouldn't that mean all the existing RayObjects would become invalid?

File: src/ray/raylet/raylet_client.h
Comment (ericl) [reply src/ray/raylet/raylet_client.h:60]: Nvm, this does go over grpc. Comment may be misleading, this method executes a grpc request exchange.

**Review by ericl — APPROVED** [review]
lgtm (only reviewed interfaces)

**Pull request merged**
